### PR TITLE
[G5] Make block_cipher work with PSA

### LINKED ISSUE #8598: [G5] Make `block_cipher` work with PSA
In the new internal abstraction layer `block_cipher`, add dispatch to PSA based on the following logic:
- if we have a driver for this cipher (according to the `ACCEL` macro) _and_ PSA is initialized -> use PSA;
- otherwise, use the software implementation.

This is similar to the logic implemented in `md.c`, see `md_can_use_psa()`, `mbedtls_md_setup()` and `mbedtls_md_update()` for example.

Add tests for this dispatch. We probably want them in a separate test suite as we did for `test_suite_md.psa.data`. Check in the output from `analyze_outcomes.py`'s coverage check that these are all executed on the CI.

Add a pair of components with the following configuration:
- both: starting from full, removing unauthenticated ciphers, but keeping AEADs (so, similar to `common_test_full_no_cipher_with_psa_crypto` with `PSA_CRYPTO_CONFIG`);
- driver: accelerated key types: AES, ARIA, Camellia - accelerated algs: ECB only.
Check for parity in `analyze_outcomes.py`.

Fix any issue or disparity found.

Add ChangeLog and update any relevant documentation.

### PR DESCRIPTION
**Pull request by valeriosetti**
This PR add the possibility to dispatch to PSA in BLOCK_CIPHER if/when possible.

Resolves #8598.

## PR checklist

- [x] **changelog** postponed to https://github.com/Mbed-TLS/mbedtls/issues/6337
- [x] **backport** not required because it's an enhancement
- [x] **tests** provided

### CODE REVIEW

File: library/block_cipher.c [hunk 1]
```diff
@@ -70,6 +122,32 @@ int mbedtls_block_cipher_setkey(mbedtls_block_cipher_context_t *ctx,
...
+    if (ctx->engine == MBEDTLS_BLOCK_CIPHER_ENGINE_PSA) {
+        psa_key_attributes_t key_attr = PSA_KEY_ATTRIBUTES_INIT;
+        psa_status_t status;
+
+        psa_set_key_type(&key_attr, ctx->psa_key_type);
+        psa_set_key_bits(&key_attr, key_bitlen);
+        psa_set_key_algorithm(&key_attr, PSA_ALG_ECB_NO_PADDING);
+        psa_set_key_usage_flags(&key_attr, PSA_KEY_USAGE_ENCRYPT);
+
+        status = psa_import_key(&key_attr, key, key_bitlen/8, &ctx->psa_key_id);
```
Comment (mpg) [inline library/block_cipher.c:135]: Minor: I think there's a macro `PSA_BITS_TO_BYTES()` which would be slightly better.

File: library/psa_util.c [hunk 2]
```diff
@@ -50,6 +53,17 @@ const mbedtls_error_pair_t psa_to_md_errors[] =
...
 };
 #endif
+
+#if defined(MBEDTLS_BLOCK_CIPHER_SOME_PSA)
+const mbedtls_error_pair_t psa_to_cipher_errors[] =
+{
+    { PSA_SUCCESS,                     0 },
+    { PSA_ERROR_NOT_SUPPORTED,         MBEDTLS_ERR_CIPHER_FEATURE_UNAVAILABLE },
+    { PSA_ERROR_INVALID_ARGUMENT,      MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA },
+    { PSA_ERROR_INSUFFICIENT_MEMORY,   MBEDTLS_ERR_CIPHER_ALLOC_FAILED }
```
Comment (mpg) [inline library/psa_util.c:63]: Note to self: in the next pass, check that this covers all error code potentially returned by the functions used.

File: include/mbedtls/config_adjust_legacy_crypto.h [hunk 3]
```diff
@@ -165,6 +165,56 @@
 
 #endif /* MBEDTLS_MD_LIGHT */
 
+/* BLOCK_CIPHER module can dispatch to PSA when:
+ * - PSA is enabled and drivers have been initialized
+ * - desired key type is supported on the PSA side
+ * If the above conditions are not met, but the legacy support is enabled, then
+ * BLOCK_CIPHER will dinamically fallback to it.
```
Comment (mpg) [inline include/mbedtls/config_adjust_legacy_crypto.h:172]: Typo: dYnamically

File: include/mbedtls/check_config.h [hunk 4]
```diff
@@ -333,13 +333,17 @@
 #endif
 #undef MBEDTLS_HAS_MEMSAN
 
-#if defined(MBEDTLS_CCM_C) && (                                        \
-    !defined(MBEDTLS_AES_C) && !defined(MBEDTLS_CAMELLIA_C) && !defined(MBEDTLS_ARIA_C) )
+#if defined(MBEDTLS_CCM_C) && \
+    !(defined(MBEDTLS_BLOCK_CIPHER_CAN_AES) || defined(MBEDTLS_AES_C) || \
```
Comment (mpg) [inline include/mbedtls/check_config.h:337]: That's not quite correct now: `CCM_C` can't benefit from AES support in `BLOCK_CIPHER_C` when `CIPHER_C` is defined, but I think that's something we should change anyway, and then this would become correct.

File: tests/scripts/all.sh [hunk 5]
```diff
@@ -3818,6 +3818,61 @@ component_test_psa_crypto_config_reference_cipher_aead () {
...
+    # Start from the full config
+    helper_libtestdriver1_adjust_config "full"
+
+    # Disable CCM/GCM support in PSA as we are testing dispatching of legacy
+    # modules (GCM_C/CCM_C through BLOCK_CIPHER_C) to PSA.
+    scripts/config.py -f $CRYPTO_CONFIG_H unset PSA_WANT_ALG_CCM
+    scripts/config.py -f $CRYPTO_CONFIG_H unset PSA_WANT_ALG_GCM
+
+    # Disable CIPHER_C because we want legacy GCM_C/CCM_C to use BLOCK_CIPHER_C.
+    scripts/config.py unset MBEDTLS_CIPHER_C
```
Comment (mpg) [inline tests/scripts/all.sh:3839]: Note: that's a usability problem. It will result in annoying documentation/ChangeLog: "Built-in GCM and CCM can now take advantage of a plain AES driver, but only if CIPHER_C is disabled. If you want those to use your driver but still need CIPHER_C for some reason (like using PKCS5/12, NIST-KW, CMAC, encrypted PEM) then tough luck."

I wonder if we should change the logic to:
- `GCM_C` & `CCM_C` will use `BLOCK_CIPHER_C` as soon as it's enabled;
- we auto-enable it if (`GCM_C` or `CCM_C` is enabled) and (`CIPHER_C` is disabled or `MBEDTLS_BLOCK_CIPHER_SOME_PSA` is defined). (Of course make the definition of `MBEDTLS_BLOCK_CIPHER_SOME_PSA` no longer conditional on `BLOCK_CIPHER_C` being enabled in the first place or we have a loop.)

Wdyt? (Could be a follow-up.)

File: tests/include/test/psa_crypto_helpers.h [hunk 6]
```diff
@@ -367,6 +367,30 @@ uint64_t mbedtls_test_parse_binary_string(data_t *bin_string);
 #define MD_PSA_DONE() ((void) 0)
 #endif /* MBEDTLS_MD_SOME_PSA */
 
+/** \def BLOCK_CIPHER_PSA_INIT
```
Comment (mpg) [inline tests/include/test/psa_crypto_helpers.h:370]: \<rant>Aw, we're really starting to have too many of these macros. Nothing you can do about it now, I just can't wait for 4.0\</rant>

**Review by mpg — CHANGES_REQUESTED** [review]
Looking pretty good! Mostly minor points.

File: library/block_cipher.c [hunk 7]
```diff
@@ -42,6 +83,17 @@ void mbedtls_block_cipher_free(mbedtls_block_cipher_context_t *ctx)
 int mbedtls_block_cipher_setup(mbedtls_block_cipher_context_t *ctx,
                                mbedtls_cipher_id_t cipher_id)
 {
+#if defined(MBEDTLS_BLOCK_CIPHER_SOME_PSA)
+    if (psa_can_do_cipher(cipher_id)) {
```
[the remark this answers is not in the document]
Comment (valeriosetti) [reply library/block_cipher.c:87]: That's due to the way I implemented it: since I was experimentally developing the feature, I first wrote all the code and then split it into commits. As a consequence it might happen that the code cannot be built in some intermediate commit. This works fine because in the end we're accepting the PR entirely or not and not just picking some commits from it. I understand that this might not be the best way to create git history for a reviewer though

File: include/mbedtls/check_config.h (hunk 4 above)
Comment (valeriosetti) [reply include/mbedtls/check_config.h:337]: As far as I see we have 2 options:
- if `CIPHER_C` is defined, then CCM/GCM can do AES (through CIPHER) as long as `AES_C` if defined
- if `CIPHER_C` is not defined, then CCM/GCM can do AES (though BLOCK_CIPHER) as long as `MBEDTLS_BLOCK_CIPHER_CAN_AES` is defined. Here `MBEDTLS_BLOCK_CIPHER_CAN_AES` strictly depends on `BLOCK_CIPHER_C`, of course,

So I think that the above line should be:
`defined(MBEDTLS_BLOCK_CIPHER_CAN_AES) || (defined(MBEDTLS_CIPHER_C) && defined(MBEDTLS_AES_C))`. Isn't it?

[I wrote about AES here, but the same holds for ARIA and Camellia.]

> I think that's something we should change anyway, and then this would become correct.

Is this related to [this comment](https://github.com/Mbed-TLS/mbedtls/pull/8632#discussion_r1425220251)?

File: tests/scripts/all.sh (hunk 5 above)
Comment (valeriosetti) [reply tests/scripts/all.sh:3839]: The main consequence I see here is that `BLOCK_CIPHER_C` and `CIPHER_C` are no more mutually exclusive: in a build which has `CIPHER_C` and there is a PSA driver available for AES keys, we end up having an increased code size. The advantage is that driver might be faster/safer for AES computations though.

Comment (valeriosetti) [reply tests/scripts/all.sh]: Commits from https://github.com/Mbed-TLS/mbedtls/pull/8632/commits/8f8a598c77d71fb71da45bd61bcc1f24a6b5619d to https://github.com/Mbed-TLS/mbedtls/pull/8632/commits/16fa1a7577fa086a3306cb44b4d0b7497cdb5a19 implement this.

**Comment by valeriosetti** [comment]
Only 1 failure in `test_no_ctr_drbg_classic` but it's related to Docker in the OpenCI. Internal CI didn't experience the same problem. I'm going to remove the `need-ci` label

**Comment by valeriosetti** [comment]
@mpg since the `analyze_coverage.py` part wasn't that big change (compared to the good amount of changes already introduced by this PR), I chose to keep everything in this PR. Let me know if that's not OK for you

**Comment by valeriosetti** [comment]
I apologize for the rebase, but there were conflicts with `development` concerning `analyze_outcomes.py`

**Comment by mpg** [comment]
> @mpg since the `analyze_coverage.py` part wasn't that big change (compared to the good amount of changes already introduced by this PR), I chose to keep everything in this PR. Let me know if that's not OK for you

Sure! I actually prefer it this way - it increases our confidence that what we're merging really behaves like we expect, and I think the size of the PR remains manageable.

File: include/mbedtls/block_cipher.h [hunk 8]
```diff
@@ -25,7 +25,7 @@
 #endif
 
 #if defined(MBEDTLS_BLOCK_CIPHER_SOME_PSA)
-#include "psa/crypto.h"
+#include "psa/crypto_types.h"
```
Comment (mpg) [inline include/mbedtls/block_cipher.h:28]: Hmm, I know I suggested this, but now I'm having doubts: this file says
```
 * \note This file may not be included directly. Applications must
 * include psa/crypto.h. Drivers must include the appropriate driver
 * header file.
```
Is the library allowed to include it directly? @gilles-peskine-arm wdyt?

For context, we're trying to avoid a dependency cycle here:
- `mbedtls_block_cipher_context_t` has fields of types `psa_key_type_t` and `mbedtls_svc_key_id_t`
- if we include `psa/crypto.h`, which is the normal way to get those types defined, then the following will be included: `crypto_struct.h` -> `crypto_driver_contexts_composites.h` -> `gcm.h` + `ccm.h`
- but there the contexts (in configs or interest) have a field of type `mbedtls_block_cipher_context_t` which is not defined because all this happened before we could reach that definition in `block_cipher.h`.

Including `crypto_types.h` here instead of `crypto.h` avoids that issue, but is it the right thing to do?

File: include/mbedtls/block_cipher.h [hunk 9]
```diff
@@ -53,7 +53,6 @@ typedef struct {
     mbedtls_block_cipher_id_t MBEDTLS_PRIVATE(id);
 #if defined(MBEDTLS_BLOCK_CIPHER_SOME_PSA)
     mbedtls_block_cipher_engine_t engine;
-    psa_cipher_operation_t psa_operation;
     psa_key_type_t psa_key_type;
```
Comment (mpg) [inline include/mbedtls/block_cipher.h]: Note: this seems somewhat redundant with `mbedtls_block_cipher_id_t` above. Could we avoid storing this one?

File: library/block_cipher.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (mpg) [reply library/block_cipher.c:86]: For review it was not a problem except for the initial surprise. Way easier than reviewing a single massive commit anyway :)

However, in general we try to structure commits so that each one at least builds, in order to facilitate bisecting should we need it in the future. But that's not a hard rule and I don't think it's worth the effort restructuring here. (Also, git bisect has provisions for skipping commits that can't be tested, it just takes a bit more care.)

File: tests/scripts/all.sh [hunk 10]
```diff
@@ -3824,6 +3837,32 @@ component_test_psa_crypto_config_reference_cipher_aead () {
...
+    msg "build: full + PSA dispatch in block_cipher"
+
+    loc_accel_list="ALG_ECB_NO_PADDING \
+                    KEY_TYPE_AES KEY_TYPE_ARIA KEY_TYPE_CAMELLIA"
+
+    # Configure
+    # ---------
+
+    # Start from the full config
+    helper_libtestdriver1_adjust_config "full"
```
Comment (mpg) [inline tests/scripts/all.sh:3850]: Hmm, it seems to me we also want to disable `AES_C`, `ARIA_C` and `CAMELLIA_C` (and test that `aes.o` etc are empty after building). After all the goal is to check that these can be completely replaced with drivers and at least a "small" config (that is, one similar to TF-M's medium profile + X.509 and TLS) should still work.

So, it's OK if we have to disable a few things like PKCS5, non-AEAD modes, etc. But we should have AES provided only by a driver and GCM+CCM still working.

As a consequence of dependencies, we might need to then disable a bit more here, and perhaps in the reference component too.

**Review by mpg — CHANGES_REQUESTED** [review]
Thanks for addressing my feedback! I have some more though :)

This is still not a complete review, mostly high-level. I'll need to make another, more in-depth pass, next week.

File: tests/scripts/all.sh (hunk 10 above)
Comment (valeriosetti) [reply tests/scripts/all.sh:3850]: I was going for keeping as many things enabled as possible, but you are right: in this was there is no "secure" way to test if block_cipher is dispatching properly or not. I will fix that

Comment (valeriosetti) [reply tests/scripts/all.sh:3850]: The main problem in doing this is that legacy modules get automatically enabled in `config_adjust_legacy_from_psa.h`. For example for AES we have:
```
#if defined(PSA_WANT_KEY_TYPE_AES)
#if !defined(MBEDTLS_PSA_ACCEL_KEY_TYPE_AES)
#define PSA_HAVE_SOFT_KEY_TYPE_AES 1
#endif /* !MBEDTLS_PSA_ACCEL_KEY_TYPE_AES */
#if defined(PSA_HAVE_SOFT_KEY_TYPE_AES) || \
    defined(PSA_HAVE_SOFT_BLOCK_MODE) || \
    defined(PSA_HAVE_SOFT_BLOCK_AEAD)
#define MBEDTLS_PSA_BUILTIN_KEY_TYPE_AES 1
#define MBEDTLS_AES_C
#endif /* PSA_HAVE_SOFT_KEY_TYPE_AES || PSA_HAVE_SOFT_BLOCK_MODE */
#endif /* PSA_WANT_KEY_TYPE_AES */
```
We are accelerating AES, but not AEADs nor ciphers (not all of them, only ECB), so we get `AES_C` enabled back.
The only solution I saw for this was to accelerate both AEADs+ciphers, but then we might have TLS using PSA driver instead of BLOCK_CIPHER and this is not what we want for this test :/

Comment (mpg) [reply tests/scripts/all.sh:3850]: The definition of G5 for this EPIC is: as a user, I have a driver that accelerates AES (Aria, Camellia) but only the primitive (that is, ECB), not any other mode; I want to be able to remove the built-in implementation of AES (Aria, Camellia) and still get enough things working to support TLS with GCM and CCM ciphersuites.

So, this `all.sh` component should:
- prove that GCM and CCM (built-in) indeed work with a driver that only accelerates ECB and built-in AES (Aria, Camellia) removed;
- clarify what else needs to be disabled / stops working when removing the built-in implementation.

File: include/mbedtls/block_cipher.h (hunk 8 above)
Comment (gilles-peskine-arm) [reply include/mbedtls/block_cipher.h:28]: Applications must not include these headers directly because we don't offer any stability guarantees. It's ok for library headers to include them. This specific case seems right to me: the reason `crypto_types.h` has declarations without definitions for struct types is precisely so that it can be included before `crypto_struct.h` is available in the header dependency chain.

Comment (mpg) [reply include/mbedtls/block_cipher.h:28]: Great, thanks!

**Comment by valeriosetti** [comment]
Note to self: is the new test component added by this PR making the already existing `component_test_psa_crypto_config_accel_cipher` (or `component_test_psa_crypto_config_accel_des` in case it will be renamed in #8358) obsolete?

**Comment by mpg** [comment]
Btw, I've made a note in #6337 to make another pass over the set of tests looking for any gaps or redundancies.

File: tests/scripts/all.sh [hunk 11]
```diff
@@ -3846,8 +3858,19 @@ component_test_full_block_cipher_psa_dispatch () {
     # Configure
     # ---------
 
-    # Start from the full config
-    helper_libtestdriver1_adjust_config "full"
+    common_block_cipher_dispatch 1
+
+    # Disable cipher's modes and AEADs that, when not accelerated, cause
```
Comment (mpg) [inline tests/scripts/all.sh:3863]: Suggestion: does putting this in the common function reduce the number of things we have to ignore in `analyze_outcomes.py`?

File: tests/scripts/all.sh [hunk 12]
```diff
@@ -3856,13 +3879,36 @@ component_test_full_block_cipher_psa_dispatch () {
...
+# This is the reference component of component_test_full_block_cipher_psa_dispatch
+component_test_full_block_cipher_legacy_dispatch () {
+    msg "build: full + legacy dispatch in block_cipher"
+
+    common_block_cipher_dispatch 0
+
+    # Disable cipher modes other than ECB as in the accelerated component. ECB
+    # does not have a configuration symbol and it's automatically enabled as
+    # long as underlying key types are.
+    scripts/config.py unset-all MBEDTLS_CIPHER_MODE
```
Comment (mpg) [inline tests/scripts/all.sh:3904]: Same suggestion: does putting this in the common function reduce the number of things we have to ignore in `analyze_outcomes.py`?

File: tests/scripts/all.sh (hunk 11 above)
Comment (valeriosetti) [reply tests/scripts/all.sh]: Yes, indeed, this is something I had locally but I didn't pushed yet ;)

File: tests/scripts/analyze_outcomes.py [hunk 13]
```diff
@@ -542,8 +542,38 @@ def do_analyze_driver_vs_reference(results: Results, outcomes: Outcomes, args) -
             'ignored_suites': [
                 # Skipped in the accelerated component
                 'aes', 'aria', 'camellia',
+                # These require AES_C and CAMELLIA_C to be enabled in order for the cipher
+                # module (actually cipher_wrapper) to work properly. However these symbols
+                # are disabled in the accelerated component so we ignore them.
```
Comment (mpg) [inline tests/scripts/analyze_outcomes.py:547]: Ok, that will be something interesting to include in the user documentation: in those builds when we say GCM and CCM work, we mean using the `gcm.h`/`ccm.h` API or the PSA Crypto API, but not the `cipher.h` API. (Nothing to change here, just making a note.)

File: tests/scripts/analyze_outcomes.py [hunk 14]
```diff
@@ -542,8 +542,38 @@ def do_analyze_driver_vs_reference(results: Results, outcomes: Outcomes, args) -
...
+                'test_suite_cipher.aria': [
+                    # Same as for test_suite_cipher.aes
+                    re.compile(r'ARIA-\d+[- ]CCM\*-NO-TAG .*'),
+                ],
+                'test_suite_cipher.camellia': [
+                    # Same as for test_suite_cipher.aes
+                    re.compile(r'CAMELLIA-\d+[- ]CCM\*-NO-TAG .*'),
+                ],
+                'test_suite_error': [
+                    # Following tests require AES_C which is disabled in the accelerated component
```
Comment (mpg) [inline tests/scripts/analyze_outcomes.py:569]: I think we already have something similar with other component. Perhaps re-use the same comment that explains how this is not really related to AES so we don't care?

**Review by mpg — COMMENTED** [review]
(Still a partial review, just did a first pass on the latest commits.)

File: include/mbedtls/check_config.h (hunk 4 above)
Comment (mpg) [reply include/mbedtls/check_config.h:337]: Yes, that was related to that comment, and I think it's correct now, so I'm marking this conversation as resolved.

File: include/mbedtls/check_config.h [hunk 15]
```diff
@@ -333,13 +333,17 @@
...
+    defined(MBEDTLS_BLOCK_CIPHER_CAN_ARIA) || defined(MBEDTLS_ARIA_C) || \
+    defined(MBEDTLS_BLOCK_CIPHER_CAN_CAMELLIA) || defined(MBEDTLS_CAMELLIA_C))
```
Comment (mpg) [inline include/mbedtls/check_config.h:339]: Note: this could be make slightly shorter by using `MBEDTLS_CCM_GCM_CAN_CAMELLIA` & Co.

File: library/psa_crypto_core.h [hunk 16]
```diff
@@ -33,6 +33,18 @@
...
+ * Tell if PSA is ready for this cipher.
+ *
+ * \note            For now, only checks the state of the driver subsystem,
+ *                  not the algorithm. Might do more in the future.
+ *
+ * \param cipher_alg  The cipher algorithm (ignored for now).
+ *
+ * \return 1 if the driver subsytem is ready, 0 otherwise.
+ */
+int psa_can_do_cipher(psa_algorithm_t cipher_alg);
```
Comment (mpg) [inline library/psa_crypto_core.h:46]: This input parameters should be key type + alg, not just alg. (Even if they're ignored for now.)

File: library/block_cipher.c [hunk 17]
```diff
@@ -42,26 +82,38 @@ void mbedtls_block_cipher_free(mbedtls_block_cipher_context_t *ctx)
...
                                mbedtls_cipher_id_t cipher_id)
 {
-    switch (cipher_id) {
+    ctx->id = (cipher_id == MBEDTLS_CIPHER_ID_AES) ? MBEDTLS_BLOCK_CIPHER_ID_AES :
+              (cipher_id == MBEDTLS_CIPHER_ID_ARIA) ? MBEDTLS_BLOCK_CIPHER_ID_ARIA :
+              (cipher_id == MBEDTLS_CIPHER_ID_CAMELLIA) ? MBEDTLS_BLOCK_CIPHER_ID_CAMELLIA :
+              MBEDTLS_BLOCK_CIPHER_ID_NONE;
+
+#if defined(MBEDTLS_BLOCK_CIPHER_SOME_PSA)
+    if (psa_can_do_cipher(cipher_id) &&
```
Comment (mpg) [inline library/block_cipher.c:91]: This is wrong, and it only compiles because C's type system is so weak: you're passing a `mbedtls_cipher_id_t` where (currently) a `psa_algorithm_t` is expected. I think we want to first get the PSA key type, then call this with `key_type, ALG_ECB_NO_PADDING` (see my comment about the signature of this function).

File: tests/scripts/all.sh [hunk 18]
```diff
@@ -3824,6 +3824,94 @@ component_test_psa_crypto_config_reference_cipher_aead () {
...
+    make test
+}
+
+# This is the reference component of component_test_full_block_cipher_psa_dispatch
+component_test_full_block_cipher_legacy_dispatch () {
+    msg "build: full + legacy dispatch in block_cipher"
+
+    common_block_cipher_dispatch 0
+
+    # Disable cipher modes other than ECB as in the accelerated component. ECB
```
Comment (mpg) [inline tests/scripts/all.sh:3904]: I don't understand why we need this here, while we don't need it in the accelerated component. I'm not saying it's wrong, but I'd like to understand.

File: tests/suites/test_suite_block_cipher.psa.data [hunk 19]
```diff
@@ -0,0 +1,39 @@
+# These tests behave differently depending on the presence of
+# drivers and/or built-in, so they're isolated here for the benefit of
+# analyze_outcomes.py (driver vs reference comparison).
```
Comment (mpg) [inline tests/suites/test_suite_block_cipher.psa.data:3]: Yet I'm not seeing `block_cipher.psa` in the list of ignored suites for this pair of components. I guess that's because in the reference component, `CIPHER_C` is enabled, and nothing's accelerated, so `BLOCK_CIPHER_C` is disabled, so the whole `test_suite_block_cipher` is skipped in the ref component, hence doesn't need an ignore entry.

Perhaps just remove the end of the sentence then (after the coma)?

**Review by mpg — CHANGES_REQUESTED** [review]
Ok, this is finally a complete review. Looking pretty good, mostly minor stuff.

File: tests/scripts/all.sh (hunk 18 above)
Comment (valeriosetti) [reply tests/scripts/all.sh]: Right, this looks more a left over from the development phase rather than something that is really required.

File: tests/scripts/analyze_outcomes.py [hunk 20]
```diff
@@ -542,28 +542,31 @@ def do_analyze_driver_vs_reference(results: Results, outcomes: Outcomes, args) -
...
                 'aes', 'aria', 'camellia',
-                # These require AES_C and CAMELLIA_C to be enabled in order for the cipher
-                # module (actually cipher_wrapper) to work properly. However these symbols
-                # are disabled in the accelerated component so we ignore them.
-                'cipher.ccm', 'cipher.gcm', 'cmac',
+                # These require AES_C, ARIA_C or CAMELLIA_C to be enabled in
+                # order for the cipher module (actually cipher_wrapper) to work
+                # properly. However these symbols are disabled in the accelerated
+                # component so we ignore them.
+                'cipher.ccm', 'cipher.gcm', 'cmac', 'cipher.aes', 'cipher.aria',
```
Comment (mpg) [inline tests/scripts/analyze_outcomes.py:549]: Ok, so here I'm going to make the comment I originally made on another PR, but this time hopefully it will make sense :) We have built-in DES, built-in CMAC, and no AES. So we shouldn't have to ignore the entire CMAC test suite: only the cases that use AES (those that use DES should work).

I suggest renaming test cases in `test_suite_cmac.data` so that all test using AES have AES in their name, then below we can just do a regex-ignore based on AES.

File: tests/scripts/all.sh [hunk 21]
```diff
@@ -3824,6 +3824,88 @@ component_test_psa_crypto_config_reference_cipher_aead () {
...
+    # Keep this also in the reference component in order to skip the same tests
+    # that were skipped in the accelerated one.
+    scripts/config.py -f "$CRYPTO_CONFIG_H" unset PSA_WANT_ALG_CTR
+    scripts/config.py -f "$CRYPTO_CONFIG_H" unset PSA_WANT_ALG_CFB
+    scripts/config.py -f "$CRYPTO_CONFIG_H" unset PSA_WANT_ALG_OFB
+    scripts/config.py -f "$CRYPTO_CONFIG_H" unset PSA_WANT_ALG_CBC_NO_PADDING
+    scripts/config.py -f "$CRYPTO_CONFIG_H" unset PSA_WANT_ALG_CBC_PKCS7
+    scripts/config.py -f "$CRYPTO_CONFIG_H" unset PSA_WANT_ALG_CMAC
+    scripts/config.py -f "$CRYPTO_CONFIG_H" unset PSA_WANT_ALG_CCM_STAR_NO_TAG
+    scripts/config.py -f "$CRYPTO_CONFIG_H" unset PSA_WANT_ALG_CCM
```
Comment (mpg) [inline tests/scripts/all.sh:3852]: Those two lines are very wrong: the whole point is that we want CCM and GCM to work here!

File: tests/scripts/analyze_outcomes.py (hunk 20 above)
Comment (mpg) [reply tests/scripts/analyze_outcomes.py:549]: Also, regarding `cipher.aes`, `cipher.gcm`, `cipher.ccm` - they have "PSA" test cases which I expect should be working here. I suggest we create a new test suite `cipher.psa.data` and move all the "PSA" test cases from the above three to the new test suites.

If I'm not mistaken, we should not have to ignore `cipher.psa` here.

**Review by mpg — CHANGES_REQUESTED** [review]
Thanks for addressing y comments. Two minor points about ignored tests, but also one big thing I missed in my previous review - hopefully it wasn't hiding to many issues :crossed_fingers:

File: tests/scripts/all.sh (hunk 21 above)
Comment (valeriosetti) [reply tests/scripts/all.sh:3852]: Yes, it is true that we want CCM/GCM to work here, but this should happen through the legacy `cipher_block` interface; we're not checking the PSA path here. Moreover, as the comment above this block explains, enabling any of these symbols without acceleration will cause legacy key types (`MBEDTLS_(AES|ARIA|CAMELLLIA)_C`) to be re-enabled in `config_adjust_legacy_from_psa.h`.
So are you suggesting to accelerate also CCM and GCM here?

File: tests/scripts/analyze_outcomes.py (hunk 20 above)
Comment (valeriosetti) [reply tests/scripts/analyze_outcomes.py:549]: > I suggest renaming test cases in `test_suite_cmac.data` so that all test using AES have AES in their name, then below we can just do a regex-ignore based on AES.

This is something I did initially, but then I realized that skipping tests with AES, ARIA and CAMELLIA (because also these 2 are used there as check) then basically only 2 items remained, so I eventually skipped the entire suite :)

File: tests/scripts/all.sh (hunk 21 above)
Comment (mpg) [reply tests/scripts/all.sh:3852]: > Edit: on a 2nd thought I think that I should update `config_adjust_legacy_from_psa.h` because even in case of builtin CCM/GCM we can rely on PSA dispatching of the block_cipher module. So please ignore the above comment for now while I will try to implement what you proposed.

Ah great, that's what I was going to say! The whole point of G5 is to be able to enable `PSA_WANT_ALG_[CG]CM` without `AES_C` (Aria, Camellia) as long as we have `ACCEL_KEY_TYPE_AES` (Aria, Camellia) + `ACCEL_ALG_ECB_NO_PADDING`. So we shouldn't stop until that's possible and proven by tests.

File: tests/scripts/analyze_outcomes.py (hunk 20 above)
Comment (mpg) [reply tests/scripts/analyze_outcomes.py]: > This is something I did initially, but then I realized that skipping tests with AES, ARIA and CAMELLIA (because also these 2 are used there as check) then basically only 2 items remained, so I eventually skipped the entire suite :)

Ah, right, I had only checked that there were DES tests, not how many. Indeed 2 is not a lot, but I think it's slightly cleaner this way - we are now testing the assertion we're making in the documentation that CMAC works with whatever cipher is built-in.

Comment (mpg) [reply tests/scripts/analyze_outcomes.py]: > These PSA tests are actually skipped because `MBEDTLS_DEPRECATED_REMOVED` is disabled. There is a comment in `common_block_cipher_dispatch()` with which I tried to explain the reason. Is it wrong?

Ah, no that's not wrong at all, I just forgot, sorry. Let's not worry about it then.

**Review by mpg — COMMENTED** [review]
Just a note that I've reviewed the last two commits and am happy with them. I'm just holding my approval until [this comment](https://github.com/Mbed-TLS/mbedtls/pull/8632#discussion_r1438136320) is addressed - which I hope will go smoothly and not uncover deep issues.

File: library/cipher_wrap.c [hunk 22]
```diff
@@ -566,14 +566,14 @@ static const mbedtls_cipher_info_t aes_256_xts_info = {
 #endif /* MBEDTLS_CIPHER_MODE_XTS */
 #endif /* MBEDTLS_AES_C */
 
-#if defined(MBEDTLS_GCM_C) && defined(MBEDTLS_AES_C)
+#if defined(MBEDTLS_GCM_C)
```
Comment (mpg) [inline library/cipher_wrap.c:569]: Shouldn't we keep something like `MBEDTLS_CCM_GCM_CAN_AES` though?

Comment (valeriosetti) [reply library/cipher_wrap.c:569]: Previous to this commit I got undefined symbol for `ccm_aes_setkey_wrap` while trying to run `test_full_block_cipher_psa_dispatch`. This is due to the fact that the `ccm_aes_info` structure was included in that build, but `ccm_aes_setkey_wrap` was not due to `MBEDTLS_AES_C` being disabled. 
Btw I agree that I might have simplified guards too much (indeed the CI complains) so I will improve that.

Comment (valeriosetti) [reply library/cipher_wrap.c]: > Shouldn't we keep something like `MBEDTLS_CCM_GCM_CAN_AES` though?

Yes, it seems to work (see [commit](https://github.com/Mbed-TLS/mbedtls/pull/8632/commits/c1c6858bfc9a2e7a9c3062fa12e0acdb6d56fff1)) but it seems to me that it adds interdependency between cipher and block_cipher. Let's see what the CI thinks about it.

**Review by mpg — APPROVED** [review]
LGTM, thanks!

File: library/block_cipher.c [hunk 23]
```diff
@@ -70,6 +123,26 @@ int mbedtls_block_cipher_setkey(mbedtls_block_cipher_context_t *ctx,
...
+    if (ctx->engine == MBEDTLS_BLOCK_CIPHER_ENGINE_PSA) {
+        psa_key_attributes_t key_attr = PSA_KEY_ATTRIBUTES_INIT;
+        psa_status_t status;
+
+        psa_set_key_type(&key_attr, psa_key_type_from_block_cipher_id(ctx->id));
+        psa_set_key_bits(&key_attr, key_bitlen);
+        psa_set_key_algorithm(&key_attr, PSA_ALG_ECB_NO_PADDING);
+        psa_set_key_usage_flags(&key_attr, PSA_KEY_USAGE_ENCRYPT);
+
+        status = psa_import_key(&key_attr, key, PSA_BITS_TO_BYTES(key_bitlen), &ctx->psa_key_id);
```
Comment (d3zd3z) [inline library/block_cipher.c:136]: Not really an issue, but the commit that changes this shows it being previous just `key_bitlen/8`, which would return one short in the case that the key bitlen is not a multiple of 8. Probably hasn't been an issue, as it seems most common for the bitlens to be 8 bits. Regardless, this fixes is, hence this just being a comment.

**Review by d3zd3z — APPROVED** [review]
Looks good. I didn't find any remaining issues.
